File: classifieds/app/src/main/java/gparap/apps/classifieds/ui/home/HomeFragment.java
/*
 * Copyright 2026 gparap
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package gparap.apps.classifieds.ui.market.categories.home;

import android.content.res.AssetManager;
import android.graphics.drawable.Drawable;
import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.fragment.app.Fragment;
import androidx.navigation.NavController;
import androidx.navigation.Navigation;

import java.util.ArrayList;

import gparap.apps.classifieds.R;
import gparap.apps.classifieds.utils.Utils;

public class HomeFragment extends Fragment {

    private HomeViewModel mViewModel;

    public static HomeFragment newInstance() {
        return new HomeFragment();
    }

    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container,
                             @Nullable Bundle savedInstanceState) {
        return inflater.inflate(R.layout.fragment_category_home, container, false);
    }

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);

        loadHomeCategoryAssets();

        openHomeCategory(R.id.imageButton_marketCategory_Home_Appliances, "Appliances");
        openHomeCategory(R.id.imageButton_marketCategory_Home_Decor, "Decor");
        openHomeCategory(R.id.imageButton_marketCategory_Home_Furniture, "Furniture");
        openHomeCategory(R.id.imageButton_marketCategory_Home_Garden, "Garden");
        openHomeCategory(R.id.imageButton_marketCategory_Home_Kitchen, "Kitchen");
        openHomeCategory(R.id.imageButton_marketCategory_Home_Other, "Other");
    }

    private void loadHomeCategoryAssets() {
        String[] assets = new String[]{"appliances.jpg", "decor.jpg", "furniture.jpg", "garden.jpg",
                "kitchen.jpg", "other.jpg"};
        AssetManager assetManager = requireActivity().getAssets();
        String path = "market/categories/home";
        ArrayList<Drawable> drawables = Utils.getInstance().getDrawablesFromAssets(assetManager, assets, path);

        //Sets the drawables as the content of the ImageViews
        ImageView appliances = HomeFragment.this.requireView().findViewById(R.id.imageButton_marketCategory_Home_Appliances);
        appliances.setImageDrawable(drawables.get(0));
        ImageView decor = HomeFragment.this.requireView().findViewById(R.id.imageButton_marketCategory_Home_Decor);
        decor.setImageDrawable(drawables.get(1));
        ImageView furniture = HomeFragment.this.requireView().findViewById(R.id.imageButton_marketCategory_Home_Furniture);
        furniture.setImageDrawable(drawables.get(2));
        ImageView garden = HomeFragment.this.requireView().findViewById(R.id.imageButton_marketCategory_Home_Garden);
        garden.setImageDrawable(drawables.get(3));
        ImageView kitchen = HomeFragment.this.requireView().findViewById(R.id.imageButton_marketCategory_Home_Kitchen);
        kitchen.setImageDrawable(drawables.get(4));
        ImageView other = HomeFragment.this.requireView().findViewById(R.id.imageButton_marketCategory_Home_Other);
        other.setImageDrawable(drawables.get(5));
    }

    private void openHomeCategory(int imageResId, String subCategoryName) {
        ImageView imageView = HomeFragment.this.requireView().findViewById(imageResId);
        imageView.setOnClickListener(v -> {
            NavController navController = Navigation.findNavController(v);
            HomeFragmentDirections.ActionHomeFragmentToBaseFragment navAction =
                    HomeFragmentDirections.actionHomeFragmentToBaseFragment();
            navAction.setArgsMarketCategoryName("Home");
            navAction.setArgsMarketSubCategoryName(subCategoryName);
            navController.navigate(navAction);
        });
    }
}

File: recipes/app/src/main/java/gparap/apps/recipes/HomeFragment.kt
/*
 * Copyright 2023 gparap
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package gparap.apps.recipes

import android.content.Context
import android.content.Intent
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ImageView
import android.widget.ProgressBar
import androidx.fragment.app.Fragment
import androidx.lifecycle.ViewModelProvider
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import com.squareup.picasso.Picasso
import gparap.apps.recipes.adapters.RecipeAdapter
import gparap.apps.recipes.data.RecipeModel
import gparap.apps.recipes.utils.AppConstants
import gparap.apps.recipes.viewmodels.HomeViewModel

class HomeFragment : Fragment() {
    override fun onCreateView(
        inflater: LayoutInflater, container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View? {
        //inflate the layout for this fragment
        val view = inflater.inflate(R.layout.fragment_home, container, false)

        //get the SharedPreferences
        val prefsRecipeOfTheDay = this.context?.getSharedPreferences("recipe_of_the_day", Context.MODE_PRIVATE)
        val prefsWhatDayIsToday = this.context?.getSharedPreferences("what_day_is_today", Context.MODE_PRIVATE)

        //get the ViewModel of this fragment
        val viewModel = ViewModelProvider(this)[HomeViewModel::class.java]

        //consume the web service to fetch featured recipes
        viewModel.getFeaturedRecipes(prefsRecipeOfTheDay, prefsWhatDayIsToday)

        //setup the featured recipes RecyclerView with adapter
        val featureRecipes = view.findViewById<RecyclerView>(R.id.recycler_view_featured_recipes)
        featureRecipes.layoutManager = LinearLayoutManager(this.context)
        val adapter = RecipeAdapter()
        featureRecipes.adapter = adapter

        //observe the featured recipes live data
        viewModel.getFeaturedRecipesLiveData().observe(viewLifecycleOwner) {
            adapter.setRecipes(it as ArrayList<RecipeModel>)
        }

        //observe the random featured recipe live data
        var recipeModel: RecipeModel? = null
        viewModel.getRandomFeaturedRecipeLiveData().observe(viewLifecycleOwner) {
            Picasso.get()
                .load(it?.imageUrl)
                .placeholder(R.drawable.ic_image_placeholder_24)
                .into(view.findViewById<ImageView>(R.id.image_view_random_recipe))

            recipeModel = it
        }.also {
            //open recipe in its details activity
            view.findViewById<ImageView>(R.id.image_view_random_recipe).setOnClickListener {
                val intent = Intent(context, RecipeDetailsActivity::class.java)
                intent.putExtra(AppConstants.RECIPE_PARCELABLE_EXTRA, recipeModel)
                startActivity(intent)
            }
        }

        //observe the loading progress
        viewModel.getLoadingProgressVisibility().observe(viewLifecycleOwner) {
            view.findViewById<ProgressBar>(R.id.progress_home).visibility = it
        }

        //return the layout for this fragment
        return view
    }
}